### DISCUSSION TITLE
Container-aware Ghostty

### DISCUSSION THREAD

**Discussion by tristan957** [ghostty-org/ghostty#2358] [Feature Requests, Ideas]
I know this is Nix-land, but let me try to explain.

Containers are very popular for development these days. Immutable operating systems are becoming more mainstream with macOS to some extent, ostree-based systems like Fedora Silverblue, and in the future, systemd-sysext based systems. The idea behind container-based development is that you install all your dependencies in a container, clone the repo in the container, and program all within the container. Some IDEs have native support for container-based development, like GNOME Builder and VSCode.

A newish VTE-based terminal, [Ptyxis](https://gitlab.gnome.org/chergert/ptyxis) (formerly Prompt) by Christian Hergert, seems to be pioneering the UX of container development with terminals. Say you're in a container currently, creating a new tab while you're in a container will spawn the new tab in the container as well. Ghostty could have the same feature for tabs and splits. Ptyxis also has really great access to containers via its left hand side menu. It is a similar UX to Windows Terminal.

I'm not entirely sure what all would go into supporting this. Hopefully smarter people than me can provide more context, but here is at least a starting point: https://gitlab.gnome.org/chergert/ptyxis/#container-support. It seems to be based on OSC 777.

**Comment by malcolmstill** [comment]
I've had a look at what is required in https://github.com/malcolmstill/ghostty/commit/2dfbe2fae2480b315709bf580a6d5255bf65a826 (this is my first time looking at the ghostty code so things may not be in the correct place at all!)

[Screencast From 2024-12-31 18-01-18.webm](https://github.com/user-attachments/assets/1c1f065e-1f02-48a4-b8a8-3607b5a944ae)

# Notes

The changes are currently only around the OSC 777 support, nothing at all around supporting a container menu (which from `ptyxis` README seems to depend on some running agent).

Only `toolbox` (https://github.com/containers/toolbox) seems to emit the OSC 777 `container` `push` / `pop` codes. Podman and distrobox seem not to currently. With the changes on my branch what this means is that only `toolbox` containers work.

It seems `ptyxis` supports only the other runtimes where those are started explicitly from its top-left menu...if you start up a podman container from an already open tab in `ptyxis` and then ask for a new tab, it will not land you in the podman container. If, however, you open a podman container from the container menu and then create a new tab it will land you in the container. This makes sense, again, given that only `toolbox` seems to emit the OSC 777 code at the moment.

**Reply by valpackett** [reply]
WezTerm lets you do a "container menu" [in a few lines of lua](https://wezterm.org/config/lua/ExecDomain.html#example-docker-domains) (well, not live-updating, of course). The key concept it has is the concept of a "domain", as in some kind of "place where to spawn a terminal". WezTerm supports SSH, WSL, serial, multiplexer, … and now "exec" which is literally just "edit the command-to-run in some way", which can be wrapping with `podman exec`, `systemd-run` or whatever else. Then it just has a "create new tab in the same domain" mechanism, completely agnostic to what "thingy" to attach the tab to :)

The OSC 777 `container` variable is an extra mechanism for indicating that you've live-entered a container with some kind of `run`/`exec` command; and it's easy enough to wrap podman to emit those using just your shell as the Ptyxis documentation suggests. If you run

```shell
printf "\033]777;container;push;%s;podman\033\\" "mycontainername"
```

Ptyxis will interpret that correctly already.
